### PR TITLE
Shrestha/prefetch right inputs

### PR DESCRIPTION
**Pull request by shresthamalik**
Enables prefetching only the inputs that need to be prefetched
- Added a util file to move the prefetching business to a function
- NGraphPrefetchedSharedData knows which inputs it needs to copy for the encap op
- All input and output pipelined tensors are shared with Prefetch op, for simplicity
- Tensor Manager keep tracks of inputs that are not prefetched but pipelined. Required so that we can copy the non-prefetched inputs in the encap compute
- Added unit tests for tensor manager
- Added end-to-end unit test for prefetching

### CODE REVIEW

File: examples/axpy_pipelined.py [hunk 1]
```diff
@@ -64,7 +64,6 @@ def run_axpy_pipeline():
 
     with tf.Session() as sess:
         # Initialize the globals and the dataset
-        sess.run(tf.global_variables_initializer())
```
Comment (sayantan-nervana) [inline examples/axpy_pipelined.py]: This line is not needed any more?

Comment (shresthamalik) [reply examples/axpy_pipelined.py]: there are no variables in this example. not sure why it was there in the first place

File: ngraph_bridge/ngraph_utils.cc [hunk 2]
```diff
@@ -57,6 +57,21 @@ vector<int> FindComplement(const int& max_element,
...
+vector<int> FindComplement(const vector<int>& superset,
+                           const vector<int>& element_set) {
+  // max size of complement is superset
+  vector<int> complement(superset.size());
+  vector<int>::iterator it = set_difference(
+      superset.begin(), superset.begin() + superset.size(), element_set.begin(),
+      element_set.begin() + element_set.size(), complement.begin());
+  complement.resize(it - complement.begin());
+  return complement;
+}
```
Comment (sayantan-nervana) [inline ngraph_bridge/ngraph_utils.cc]: Maybe we rewrite the older `vector<int> FindComplement(const int& max_element, const vector<int>& element_set)` in terms of `vector<int> FindComplement(const vector<int>& superset, const vector<int>& element_set)`

File: test/python/test_axpy_pipelined.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 
 # For eg. when running the test from ngraph-bridge/build_cmake/test/python
 # you can add this path as below
-#sys.path.insert(0, '../../examples')
+sys.path.insert(0, '../../examples')
```
Comment (sayantan-nervana) [inline test/python/test_axpy_pipelined.py]: This is uncommented to test locally, so we can comment it back

File: ngraph_bridge/ngraph_encapsulate_op_utils.cc [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        NGraphPrefetchSharedResouce::CONTAINER_NAME,
+        NGraphPrefetchSharedResouce::RESOURCE_NAME, &shared_data);
+
+    if (!s.ok()) {
+      // We are using this for the first time i.e., we need to do the following
+      // 1. Create the shared data object
+      // 2. We get another pipelined tensor pair for the current iteration and
+      //   add it to the shared data. It will be accessed by prefetcher to copy
+      //   the
+      // prefetched inputs to device
```
Comment (kanvi-nervana) [inline ngraph_bridge/ngraph_encapsulate_op_utils.cc]: minor : this can be 1 line

File: test/test_ngraph_tensor_manager.cpp [hunk 5]
```diff
@@ -202,12 +245,24 @@ TEST_F(NGraphTensorManagerTest, NoVariablesHasPrefetch) {
   int number_of_outputs = 2;
 
   // expected
+  // var
   vector<int> empty;
+  vector<int> expected_out_indexes_need_copy = FillRange(number_of_outputs);
```
Comment (shresthamalik) [inline test/test_ngraph_tensor_manager.cpp]: The test might seem redundant. I have 4 states capturing the different combinations for existence of Variable and Prefetch. I can reduce it 2 states.

File: ngraph_bridge/ngraph_encapsulate_op.cc [hunk 6]
```diff
@@ -454,159 +458,51 @@ void NGraphEncapsulateOp::ComputeUsingParallelExecutor(OpKernelContext* ctx) {
...
+      pipelined_io_tensors;
+  OP_REQUIRES_OK(ctx, GetPipelinedIOTensorsReadyForExecution(
+                          ctx, tf_input_tensors, pipelined_tensor_store,
+                          tensor_manager, pipelined_io_tensors));
+
+  int current_iter_pipeline_depth = get<0>(pipelined_io_tensors);
   vector<shared_ptr<ng::runtime::Tensor>> ng_inputs(num_of_inputs);
   vector<shared_ptr<ng::runtime::Tensor>> ng_outputs(num_of_outputs);
 
   // Assume All inputs and outputs are pipelined
```
Comment (vishakha-nervana) [inline ngraph_bridge/ngraph_encapsulate_op.cc]: Get rid of this comment

File: ngraph_bridge/ngraph_executor.cc [hunk 7]
```diff
@@ -473,27 +477,29 @@ NGraphExecutor::InitializeIOTensorPipeline(
...
-  size_t num_inputs = ng_exec->get_parameters().size();
-  size_t num_outputs = ng_exec->get_results().size();
-
-  if (num_outputs == 0) {
-    return std::make_pair(
-        errors::Internal("Bad input/output length. Input size: ", num_inputs,
-                         " Output size: ", num_outputs),
-        nullptr);
-  }
+  size_t num_pipelined_inputs = pipelined_input_indexes.size();
```
Comment (vishakha-nervana) [inline ngraph_bridge/ngraph_executor.cc]: what is difference between ng_exec->get_parameters().size(); and pipelined_input_indexes.size(), if they both are same then why this change?

Comment (shresthamalik) [reply ngraph_bridge/ngraph_executor.cc]: They are not the same
ng_exec->get_parameters().size() returns the total number of inputs for this executable
We are interested in only the ones that are pipelined inputs

File: ngraph_bridge/ngraph_encapsulate_op.cc (hunk 6 above)
Comment (shresthamalik) [reply ngraph_bridge/ngraph_encapsulate_op.cc]: Second thoughts, this comment is still valid
All inputs and outputs are still pipelined.
Of all these pipelined inputs some are prefetched and that has been taken care of in this PR

Before this PR,
All inputs and outputs were pipelined. And all inputs were assumed to be also prefetched.

File: ngraph_bridge/ngraph_encapsulate_op_utils.cc (hunk 4 above)
Comment (shresthamalik) [reply ngraph_bridge/ngraph_encapsulate_op_utils.cc]: done

File: ngraph_bridge/ngraph_prefetch_shared_data.h [hunk 8]
```diff
@@ -115,12 +120,16 @@ class NGraphPrefetchSharedResouce : public ResourceBase {
...
   int GetSkipCount() { return m_skip_count; }
 
+  const vector<int>& GetPrefetchInputIndexes() {
+    return m_prefetch_input_indexes;
+  }
+
  private:
   const std::string m_ng_enc_op_name;
-  const std::string m_backend_name;
+  // const std::string m_backend_name;
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply ngraph_bridge/ngraph_prefetch_shared_data.h]: done

File: ngraph_bridge/ngraph_tensor_manager.h [hunk 9]
```diff
@@ -63,14 +72,26 @@ class NGraphTensorManager {
     return m_pipelined_output_indexes;
   }
 
+  // wrt to all inputs
   const vector<int>& GetPrefetchedInputIndexes() {
     return m_prefetched_input_indexes;
   }
 
+  // wrt to all inputs
+  const vector<int>& GetPipelinedButNotPrefetchedInputIndexes() {
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply ngraph_bridge/ngraph_tensor_manager.h]: Discussed in person

File: test/python/test_axpy_pipelined.py (hunk 3 above)
Comment (shresthamalik) [reply test/python/test_axpy_pipelined.py]: there might be no harm having it there and helps when running locally. What do you think? Or could it interfere when running elsewhere?

File: test/python/test_prefetched.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
+import sys
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply test/python/test_prefetched.py]: Good eye. Done!

File: test/python/test_api.py [hunk 11]
```diff
@@ -38,10 +38,6 @@ def test_enable(self):
     def test_backends_len(self):
         assert ngraph_bridge.backends_len()
 
-    def test_set_backend(self):
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply test/python/test_api.py]: This test sets backend. and does not reset it. We should not have tests setting the backend as it would affect the entire suite. The functionality has been tested in C++ tests. So we dont need it

File: ngraph_bridge/ngraph_tensor_manager.h [hunk 12]
```diff
@@ -80,17 +101,23 @@ class NGraphTensorManager {
   int m_number_of_outputs;
 
   // Book-keeping for weights-on-device optimizations
+  // indexes wrt all inputs/outputs
   vector<int> m_input_indexes_from_variables;
   vector<int> m_output_indexes_assigning_variable;
   vector<int> m_output_indexes_that_need_copy;
 
-  // All indexes that are not for from/to variables
+  // All indexes that are not from/to variables
```
[the remark this answers is not in the document]
Comment (shresthamalik) [reply ngraph_bridge/ngraph_tensor_manager.h]: I wont be able to change all the names in this PR. I can make changes to atmost  2-3 members + functions. Making changes to all can be taken in a separate PR. 

**Some Explanation**
There are only two data structures that are relative to pipelined inputs rather that all inputs. they have similar names and start with `GetPipelinedInputIndexes` to denote that we are asking relative to Pipelined Indexes (At least that's what i thought). The header also documents the indexes are wrt to what. May be this will help and is sufficient.

GetPipelinedInputIndexesThatArePrefetched() / m_pipelined_input_indexes_prefetched
GetPipelinedInputIndexesThatAreNotPrefetched() / m_pipelined_input_indexes_not_prefetched

File: ngraph_bridge/ngraph_encapsulate_op.cc [hunk 13]
```diff
@@ -454,159 +458,51 @@ void NGraphEncapsulateOp::ComputeUsingParallelExecutor(OpKernelContext* ctx) {
...
   // TODO: Fit in variables
-  ng_inputs = get<1>(io_tensors);
```
Comment (sayantan-nervana) [inline ngraph_bridge/ngraph_encapsulate_op.cc]: This is a good idea moving this chunk of code to a new file 👍

File: ngraph_bridge/ngraph_encapsulate_op_utils.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "logging/ngraph_log.h"
+#include "ngraph_bridge/ngraph_pipelined_tensors.h"
+#include "ngraph_bridge/ngraph_tensor_manager.h"
+
+using namespace std;
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
```
Comment (sayantan-nervana) [inline ngraph_bridge/ngraph_encapsulate_op_utils.h]: Please add some comments here about what this function does

File: ngraph_bridge/ngraph_prefetch_shared_data.h [hunk 15]
```diff
@@ -39,20 +39,19 @@ namespace ngraph_bridge {
 class NGraphPrefetchSharedResouce : public ResourceBase {
  public:
   explicit NGraphPrefetchSharedResouce(const std::string& ng_enc_op_name,
-                                       const std::string& backend_name,
-                                       int cluster_id, int graph_id)
+                                       int cluster_id, int graph_id,
+                                       const vector<int> prefetch_input_indexes)
```
Comment (sayantan-nervana) [inline ngraph_bridge/ngraph_prefetch_shared_data.h]: maybe you meant `const vector<int>&`?

File: ngraph_bridge/ngraph_tensor_manager.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (kanvi-nervana) [reply ngraph_bridge/ngraph_tensor_manager.h]: It might not be as straight forward to understand for a person reading the code for the first time. I would still suggest to change whatever you can in this PR to something more descriptive.

[the remark this answers is not in the document]
Comment (shresthamalik) [reply ngraph_bridge/ngraph_tensor_manager.h]: I can change these two and their corresponding functions. Will wait for others suggestions before making the change
m_prefetched_in_indxs_from_pipelined_ins
m_not_prefetched_in_indxs_from_pipelined_ins

File: ngraph_bridge/ngraph_encapsulate_op.cc (hunk 6 above)
Comment (vishakha-nervana) [reply ngraph_bridge/ngraph_encapsulate_op.cc]: In that case Get rid of word 'Assume',  and mention what you just wrote in the comment.

File: test/python/test_prefetched.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        pl1 = tf.placeholder(dtype=dtypes.int32)
+        pl1_f = tf.cast(pl1, tf.float32)
+        pl2 = tf.placeholder(dtype=dtypes.int32)
+        pl2_f = tf.cast(pl2, tf.float32)
+
+        mul = tf.compat.v1.math.multiply(pl2_f, input_f)
+        add = tf.compat.v1.math.add(mul, c2)
+        add2 = add + pl1_f * c1
+        output = add2
+        return output, pl1, pl2
```
Comment (sayantan-nervana) [inline test/python/test_prefetched.py]: Is there any difference in build_model1 and build_model2, other than that test_prefetch1 hangs on setting NGRAPH_TF_DISABLE_DEASSIGN_CLUSTERS

What I mean is, is there any particular graph structure you are testing with model 1 vs model 2? If so can you please add a comment

File: test/python/test_prefetched.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        sess.run(iterator.initializer)
+
+        for i in range(1, 10):
+            output = sess.run(model, feed_dict={pl1: i, pl2: i + 3})
+            outputs.append(output)
+
+        return outputs
+
+    # test hangs when "NGRAPH_TF_DISABLE_DEASSIGN_CLUSTERS" is set
+    def test_prefetch1(self):
```
Comment (sayantan-nervana) [inline test/python/test_prefetched.py]: Also, test_prefetch1 and test_prefetch2 are very similar. Maybe we can package it into a helper function and test_prefetch1 calls helper with build_model1 and test_prefetch2 calls helper with build_model2

File: test/test_ngraph_tensor_manager.cpp [hunk 18]
```diff
@@ -107,25 +107,42 @@ TEST_F(NGraphTensorManagerTest, NoVariablesNoPrefetch) {
...
+                                     ng_encap_graph_id, number_of_inputs,
+                                     number_of_outputs);
 
   // var related
   ASSERT_EQ(empty, tensor_manager.GetInputIndexesFedByVariables());
   ASSERT_EQ(empty, tensor_manager.GetOutputIndexesAssigningVariables());
-  ASSERT_EQ(empty, tensor_manager.GetOutputIndexesThatNeedCopy());
+  ASSERT_EQ(expected_out_indexes_need_copy,
+            tensor_manager.GetOutputIndexesThatNeedCopy());
+  // piplined
```
Comment (kanvi-nervana) [inline test/test_ngraph_tensor_manager.cpp]:
```suggestion
  // pipelined
```

File: ngraph_bridge/ngraph_tensor_manager.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (kanvi-nervana) [reply ngraph_bridge/ngraph_tensor_manager.h]: Another suggestion:
GetPipelinedInputIndexesThatArePrefetched() / m_pipelined_input_indexes_that_are_prefetched
GetPipelinedInputIndexesThatAreNotPrefetched() / m_pipelined_input_indexes_that_are_not_prefetched

File: ngraph_bridge/ngraph_encapsulate_op_utils.h (hunk 14 above)
Comment (shresthamalik) [reply ngraph_bridge/ngraph_encapsulate_op_utils.h]: done

File: test/python/test_prefetched.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        pl2 = tf.placeholder(dtype=dtypes.int32)
+        pl2_f = tf.cast(pl2, tf.float32)
+
+        mul = tf.compat.v1.math.multiply(input_f, c1)
+        add = tf.compat.v1.math.add(mul, pl2_f)
+        add2 = add + pl1_f
+        output = add2 - c2
+        return output, pl1, pl2
+
+    def build_model2(self, input_array, c1, c2):
```
Comment (shresthamalik) [inline test/python/test_prefetched.py]: I was trying to play with the input indexes. ie in model1 input index0 is prefetched and model2 index 1 is prefetched. i just dumped the graphs and TF always creates the graph with index 0 as prefetched.  I will get rid of model 2/test2

File: test/python/test_prefetched.py (hunk 16 above)
Comment (shresthamalik) [reply test/python/test_prefetched.py]: I was trying to play with the input indexes. ie in model1 input index0 is prefetched and model2 index 1 is prefetched. i just dumped the graphs and TF always creates the graph with index 0 as prefetched. I will get rid of model 2/test2. 

Tried to move stuff around but cant get the indexes to change
